Make it possible to install the static library

remi-moyen opened the pull request.

remi-moyen committed:
Fix definition of includes for gmtsph, the status (public/private/interface) was not correctly passed when and this caused an error if trying to install those files.

remi-moyen committed:
Add optional installation of static library.

I tried to make it so that "make install" does not depend on "static" (to avoid compiling the static library when this wasn't really asked), by excluding some bits from "static". This means to install the static library (if compiled), it must be explicitly required with "cmake --install /path/to/build/dir --component static".

The main drawback is that this only works on Linux because on Windows the ARCHIVE component (which contains the static library) also contains the DLL export files (which is needed if other code wants to link with gstlearn).

remi-moyen committed:
Silence messages about copying the includes when running "make install".

They take a lot of output and make it harder to see if the other files (libraries) are correctly copied. Plus they don't really have any possibility to fail, or to only copy bits of the includes, so they don't add much information.

remi-moyen commented on cmake/install.cmake:
```diff
@@ -28,19 +28,53 @@ target_include_directories(shared PUBLIC
+    ARCHIVE COMPONENT static EXCLUDE_FROM_ALL
+  )
+  install(
+    DIRECTORY ${NLopt_DIR}/../../../
```
Yes, the main issue is that NLopt is really an external dependency for gstlearn. So the easy answer it to tell users of gstlearn to go and fetch that dependency themselves, but it's annoying to have that additional dependency.

But since it only really needs the static library (and when building gstlearn as a shared library, if NLopt is static, an user of gstlearn doesn't even see the dependency to NLopt as it's hidden inside gstlearn, like for gmtsph), I feel that it should be possible to hide that.

At the moment I don't get the proper dependency relationship in gstlearn CMake files, but that may be because my NLopt is static (though even in this case, this is an issue, since if I forget to manually add the dependency to NLopt CMake doesn't complain but I get missing symbols errors later...). I haven't tried with a dynamic one.

I'm trying various things to try and help in that regard, nothing yet fully satisfactory. Maybe you're right and I'll have to add the dependency on NLopt to my user code. I don't like it, especially since I only need that when using a static gstlearn (again, the shared one hides the dependency when NLopt itself is built as static)...

remi-moyen replied:
I think the best solution is indeed to keep NLopt as an external dependency and thus not install it.

But this requires changing the `target_link_libraries()` to use `PUBLIC` when building the `static` target to properly set this dependency. We can also set it to `PUBLIC` when building `shared`, if NLopt itself is a shared library (currently in this case the link is still `PRIVATE` so the installed gstlearn is not seen as dependent on NLopt, but it actually is!).

So the part to link with NLopt would become:
```
# Link to NLopt
set(nlopt_link PUBLIC)
get_target_property(nlopt_lib_type NLopt::nlopt TYPE)
if (${FLAVOR} MATCHES shared AND ${nlopt_lib_type} MATCHES STATIC_LIBRARY)
  set(nlopt_link PRIVATE)
endif()
target_link_libraries(${FLAVOR} ${nlopt_link} NLopt::nlopt)
```

(I tried using a generator expression:
`$<IF:$<AND:$<STREQUAL:$<TARGET_PROPERTY:NLopt::nlopt,TYPE>,STATIC_LIBRARY>,$<STREQUAL:${FLAVOR},"shared">>,PRIVATE,PUBLIC>`
but this doesn't work, either because I messed up the generator expression (very likely!) or because we can't use a generator expression to set the `PUBLIC`/`PRIVATE` link type? In any case I feel it's way less readable than breaking it over a few more lines...)

Following the same logic, I believe that `gmtsph` should be set as a `PUBLIC` link when building the `static` target. The target is provided by gstlearn itself so it shouldn't change much to users of gstlearn, but it should avoid having to explicitly link with `gmtsph` when linking with the static gstlearn.

fabien-ors replied:
Sounds good for me. You are right.

gmtsph is compiled as a library because the external code is LGPL.
I am not sure that we can deploy it as an object library...

@pierre-guillou Thank you very much for having initiated this fruitful discussion!

remi-moyen committed:
Set dependency on NLopt to PUBLIC or PRIVATE (instead of always PRIVATE) depending on how NLopt was built (static/dynamic) and the target (shared/static) being built.

remi-moyen committed:
Remove installation of NLopt when installing the static target.

The previous version was assuming that NLopt had been built in a specific location but it can come from any source (e.g., a system package). Moreover, as it is just an external dependency, installing it as part of this package is weird, so it's best to leave it as an external thing.

Also updated some comments about dependencies (gmtsph and NLopt).

remi-moyen committed:
Make the installed package depend on NLopt if needed.

The conditional dependency is correct for users of gstlearn::shared but not for gstlearn::static, because that last one is not always built/installed. Users of gstlearn::static will get missing package errors and have to manually do a find_package(NLopt) (and OpenMP and Boost).

remi-moyen committed:
Merge pull request #655 from gstlearn/dev

Rebase branch for PR #631

remi-moyen committed:
Use a PUBLIC dependency on gmtsph for gstlearn::static (and keep it PRIVATE for gstlearn::shared).

remi-moyen commented:
I have tried to implement all that we discussed. It is a bit wonky because I have kept the whole "install the static library" optional, so some dependencies that are required for the static library are not perfectly handled. I couldn't find a better way to do it.

To sum up:
- `gstlearn::static` is still an optional install target that must be triggered manually, to keep it optional.
- NLopt is no longer installed when doing `make install` (as in my first PR). It is now assumed to be an entirely external dependency (if needed, see below), like OpenMP or Boost.
- If NLopt (when building gstlearn) is a dynamic library, there is now a `PUBLIC` dependency on it (both for the targets themselves and for a `find_dependency()` call). If NLopt is a static library, there is no `find_dependency()` (so like before) but `gstlearn::static` still has it in its `PUBLIC` dependencies while it is `PRIVATE` for `gstlearn::shared`. So users of `gstlearn::shared` don't need to care but users of `gstlearn::static` will need to provide NLopt by themselves. That last part is a bit annoying but I couldn't find a way around it.
- gmtsph is now also a `PUBLIC` dependency for `gstlearn::static` (but not `gstlearn::shared`), for the same reasons. Though in this case since gmtsph is entirely handled internally by gstlearn itself, there is no packaging issue, so that part works fine.

If anyone knows how to have a `find_dependency()` that is only done when a specific target is *used* (i.e., when a user does `target_link_libraries(foo gstlearn::static)`), or some other hack to achieve the same result, we could fix the ugliness, but I didn't find a way. I can imagine 2 solutions (neither are nice): we could make 2 separate gstlearn packages, or we could always force the installation of `gstlearn::static` (which means always compiling it i.e., doubling the compilation time/size!).

Note that the same issue happens with OpenMP and Boost, both of which are needed for `gstlearn::static` but not for `gstlearn::shared` so a solution for NLopt would fix those as well.

remi-moyen committed:
Make dependency on NLopt and gmtsph always PRIVATE, even for gstlearn::static.

The static library still depends on those but this is actually correctly handled by CMake (in the installed package, the INTERFACE_LINK_LIBRARIES uses $<LINK_ONLY:...> for those so that there is a dependency on the actual libraries but without pulling in the includes, defines etc.). The previous commit that made the dependency PUBLIC was therefore wrong.

Also updated a related comment in install.cmake.

fabien-ors commented:
> This doesn't change how `find_dependency()` is done (for NLopt) but I think the way CMake packaging works is that it is expected that consumers of `gstlearn::static` will do all the `find_package()` themselves (for OpenMP and Boost, but also in theory for NLopt). I don't like it (to me it should be done by the package that needs it i.e., by gstlearn itself) but this seems to be the standard CMake way. 
Yes this is the standard way (even without CMake) : Consumers of a static librarie must add all link options even those required by the static library itself.
Thank you for your investigation.

fabien-ors reviewed (approved):
Thx Rémi.
I am not sure to measure the impact of this PR on the installation (and the usage of the installed library). We should add a non-regression test for "install" target in the future.
